awscli/customizations/s3/s3handler.py

```diff
@@ -513,3 +521,197 @@ def _process_transfer(self, future):
+
+    def _do_submit(self, fileinfo):
+        extra_args = {}
+        self.REQUEST_MAPPER_METHOD(extra_args, self._cli_params)
+        subscribers = []
+        self._add_additional_subscribers(subscribers, fileinfo)
+        # The result subscriber class should always be the last registered
+        # subscriber to ensure it is not missing any information that
+        # may have been added in a different subscriber such as size.
+        subscribers.append(self.RESULT_SUBSCRIBER_CLASS(self._result_queue))
+        self._submit_transfer_request(fileinfo, extra_args, subscribers)
+
+    def _add_additional_subscribers(self, subscribers, fileinfo):
+        pass
+
+    def _submit_transfer_request(self, fileinfo, extra_args, subscribers):
+        raise NotImplementedError('_submit_transfer_request()')
+
+    def _warn_and_signal_if_skip(self, fileinfo):
+        for warning_handler in self._get_warning_handlers():
+            if warning_handler(fileinfo):
+                # On the first warning handler that returns a signal to skip
+                # immediately propogate this signal and no longer check
+                # the other warning handlers as no matter what the file will
+                # be skipped.
+                return True
+
+    def _get_warning_handlers(self):
+        # Returns a list of warning handlers, which are callables that
+        # take in a single parameter representing a FileInfo. It will then
+        # add a warning to result_queue if needed and return True if
+        # that FileInfo should be skipped.
+        return []
+
+    def _should_inject_content_type(self):
+        return (
+            self._cli_params.get('guess_mime_type') and
+            not self._cli_params.get('content_type')
+        )
+
+    def _warn_glacier(self, fileinfo):
+        if not self._cli_params.get('force_glacier_transfer'):
+            if not fileinfo.is_glacier_compatible():
+                LOGGER.debug(
+                    'Encountered glacier object s3://%s. Not performing '
+                    '%s on object.' % (fileinfo.src, fileinfo.operation_name))
+                if not self._cli_params.get('ignore_glacier_warnings'):
+                    warning = create_warning(
+                        's3://'+fileinfo.src,
+                        'Object is of storage class GLACIER. Unable to '
+                        'perform %s operations on GLACIER objects. You must '
+                        'restore the object to be able to the perform '
+                        'operation.' %
+                        fileinfo.operation_name
+                    )
+                    self._result_queue.put(warning)
+                return True
+        return False
+
+
+class UploadRequestSubmitter(BaseTransferRequestSubmitter):
+    REQUEST_MAPPER_METHOD = RequestParamsMapper.map_put_object_params
+    RESULT_SUBSCRIBER_CLASS = UploadResultSubscriber
+
+    def is_valid_submission(self, fileinfo):
```

Comment:
Yeah the O(n) I'm not too worried about because we shouldn't have too many of these (though I suppose it is wasteful to call `is_valid_submission` for (worst case) every possible submitter).  It's more the subtleties in the required ordering of submitters in order to ensure correct behavior.

FWIW, for streams would we not just have separate handlers (and therefore a separate dict of operation_name: SubmitterCls) for streams?  Not sure how you envision streams integrating, but I would assume we'd decide if we're in stream mode as high up the stack as possible and create the appropriate set of "stream" objects to use.  At that point, the rest of the code would all work the same.